# Homework

## 1. Getting pandas version

In [3]:
import pandas as pd
import numpy as np
pd.__version__

'2.2.2'

## 2. Getting the data

In [5]:
https = 'http://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv'

In [6]:
# !wget $https

In [7]:
df = pd.read_csv('laptops.csv')
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

## 3. Record counts 

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2160 entries, 0 to 2159
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Laptop        2160 non-null   object 
 1   Status        2160 non-null   object 
 2   Brand         2160 non-null   object 
 3   Model         2160 non-null   object 
 4   CPU           2160 non-null   object 
 5   RAM           2160 non-null   int64  
 6   Storage       2160 non-null   int64  
 7   Storage type  2118 non-null   object 
 8   GPU           789 non-null    object 
 9   Screen        2156 non-null   float64
 10  Touch         2160 non-null   object 
 11  Final Price   2160 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 202.6+ KB


The data inside is 2160

## 4. Laptop brands

In [12]:
df.Brand.nunique()

27

Total brands in dataset is 27

## 5. Missing values

In [15]:
df.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

There are 2 columns, `Storage type` and `GPU`

## 6. Maximum final price

In [18]:
df[df['Brand'] == 'Dell'].describe()

RAM      Storage     Screen  Final Price
count  84.000000    84.000000  84.000000    84.000000
mean   11.380952   395.142857  14.686905  1153.839881
std     4.583640   175.543846   1.023669   671.795071
min     4.000000   128.000000  13.000000   379.000000
25%     8.000000   256.000000  14.000000   699.000000
50%     8.000000   256.000000  14.500000  1003.000000
75%    16.000000   512.000000  15.600000  1313.810000
max    32.000000  1000.000000  17.000000  3936.000000

The max final price for laptop brand Dell is 3936

## 7. Medium value of Screen

In [21]:
# checking the median for Screen
df.Screen.median()

15.6

In [22]:
df.Screen.mode()

0    15.6
Name: Screen, dtype: float64

In [23]:
# filling missing value with most frequent data
df.Screen = df.Screen.fillna(df.Screen.mode()[0])
df.Screen.median()

15.6

In [24]:
df.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             0
Touch              0
Final Price        0
dtype: int64

There is no change, cause there's no missing value in screen. So the median is not changing 

## 8. Sum of weights

In [27]:
df_injoo = df[df.Brand == 'Innjoo']

In [28]:
columns = ['RAM', 'Storage', 'Screen']
df_training = df_injoo[columns]
df_training

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [46]:
X = np.array(df_training)
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [48]:
XTX = X.T.dot(X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [54]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [52]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])
y

array([1100, 1300,  800,  900, 1000, 1100])

In [60]:
w = XTX_inv.dot(X.T).dot(y)
w.sum()

91.29988062995815

The sum is 91.30